### PR TITLE
DM-14728: add asEigen function to return Eigen::Map

### PR DESCRIPTION
**Pull request by r-owen**
Also make building EigenView optional
Optionally include `pybind11/eigen.h` in `ndarray/eigen.h` (this is not compatible with EigenView)
Remove SWIG support
Fix a build issue (m4 might not be run in the right order on header files)

### CODE REVIEW

File: include/ndarray/eigen.h [hunk 1]
```diff
@@ -26,407 +26,241 @@
...
+ *  (Matrix/Array) and the compile-time dimensions.
+ *
+ *  Convenience functions `asEigenArray` and `asEigenMatrix`
+ *  return the stated expression type.
+ *
+ *  Be careful not call these functions on a temporary ndarray array,
+ *  because the memory will likely go away before you have used the returned Eigen map.
+ *  For example the following will have undefined behavior:
+ *
+ *     processMatrix(asEigenMatrix(makeNdArray(...)));  // broken
```
[the remark this answers is not in the document]
Comment (r-owen) [reply include/ndarray/eigen.h]: Thanks. I've overhauled that.

File: include/ndarray/eigen.h [hunk 2]
```diff
@@ -26,407 +26,241 @@
...
-};
+//@{
+/**
+ *  @name Eigen3 Interface
+ *
+ *  Return an `Eigen::Map` view to an ndarray array,
+ *  whose memory is owned by the ndarray array.
+ *
+ *  Template parameters optionally control the expression type
+ *  (Matrix/Array) and the compile-time dimensions.
```
[the remark this answers is not in the document]
Comment (r-owen) [reply include/ndarray/eigen.h]: Could you please suggest better wording? I've never understood why the row and column parameters exist. I just copied that doc from ArrayBase.h's docs for ndarray::ArrayBase::asEigen:

    /**
     *  @name Eigen3 Interface
     *
     *  These methods return Eigen3 views to the array.  Template
     *  parameters optionally control the expression type (Matrix/Array) and
     *  the compile-time dimensions.
     *
     *  The inline implementation is included by ndarray/eigen.h.
     */

[the remark this answers is not in the document]
Comment (TallJimbo) [reply include/ndarray/eigen.h]: Yes, this a difference from the old `asEigen` method.

How about just:
```
Use asEigen<Eigen::ArrayXpr>(array) or asEigenArray(array) to obtain an Eigen::Array-like view, and asEigen<Eigen::MatrixXpr>(array) or asEigenMatrix(array) to obtain an Eigen::Matrix-like view.
```

**Comment by r-owen** [comment]
Unfortunately `converter.h` is still used. Removing it (and the one import in `ndarray/converter/eigen.h`) results in build errors such as the following when building EigenView:
```
/Users/rowen/UW/LSST/ndarray_github/include/ndarray/EigenView.h:287:12: error: unknown type name 'Index'; did you mean 'Eigen::Index'?
    inline Index innerStride() const { return ST::getInnerStride(*Access::getCore(_array)); }
           ^~~~~
           Eigen::Index
```
I didn't even try removing the converter/*.h files. I don't doubt a lot of that can go away, but I'm not sure exactly what.

File: include/ndarray/eigen.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (r-owen) [reply include/ndarray/eigen.h]: Thanks! Adopted with minor tweaks

**Comment by TallJimbo** [comment]
That's concerning; `converter.h` is only for Python bindings (it includes `Python.h`), so a header like `EigenView.h` really shouldn't have a dependency on it.  I think there must be some implicit includes causing problems.  I've got a few minutes now; I'll see if I can build your branch and reproduce.

**Comment by r-owen** [comment]
I think common code may have snuck into `converter.h` and/or `converter/*.h`. For example `makePyManager` is used in `pybind11.h`. I'm all for pruning dead code but I think it'll be a bit tricky.

**Comment by r-owen** [comment]
Well done stripping the old code! I think my attempts were bogus because I was accidentally building against Eigen 3.3.

I have pushed another change: make pybind11_test.py only run Eigen matrix tests if built with EigenView support (restoring some test code I stripped earlier). I also supplemented the Travis script to get it to build 3 configurations; we'll see if that works.

**Comment by TallJimbo** [comment]
r.e. the last commit: the bug in the CMake build is "just" the fact that the outputs aren't restricted to the build directory.  If they were, we could have different build directories for the different configurations we want to test, and they wouldn't depend on each other at all.  So while the current situation is far from ideal, it's not something new.

AFAIK, you've addressed all of the concerns about this branch we discussed before, and I'm happy to merge it.  Is there anything else I'm forgetting?

**Pull request merged**